src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -1079,18 +1087,96 @@ public String Replace(char oldChar, char newChar)
             }
         }
 
-        // This method contains the same functionality as StringBuilder Replace. The only difference is that
-        // a new String has to be allocated since Strings are immutable
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        private extern String ReplaceInternal(String oldValue, String newValue);
-
-        public String Replace(String oldValue, String newValue)
+        public string Replace(string oldValue, string newValue)
         {
             if (oldValue == null)
                 throw new ArgumentNullException(nameof(oldValue));
-            // Note that if newValue is null, we treat it like String.Empty.
+            if (oldValue.Length == 0)
+                throw new ArgumentException(SR.Argument_StringZeroLength, nameof(oldValue));
+
```

Comment:
bail out early if oldValue same as newValue - at least Reference.equals anyway?

Strictly a breaking change as you no longer get a new instance.

Reply:
It is not clear whether this optimization would kick in frequently enough to pay for itself. We have a number of discussions about similar special cases in various PRs.